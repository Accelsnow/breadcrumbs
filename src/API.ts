import type { MultiGraph } from "graphology";
import type { App } from "obsidian";
import { ARROW_DIRECTIONS, DIRECTIONS } from "./constants";
import type { BCAPII, Directions, UserHier } from "./interfaces";
import type BCPlugin from "./main";
import { getMatrixNeighbours } from "./Views/MatrixView";
import {
  buildObsGraph,
  dfsAllPaths,
  getSubForFields,
  getSubInDirs,
} from "./Utils/graphUtils";
import {
  getFieldInfo,
  getFields,
  getOppDir,
  getOppFields,
  iterateHiers,
} from "./Utils/HierUtils";
import { createIndex } from "./Commands/CreateIndex";

export class BCAPI implements BCAPII {
  app: App;
  plugin: BCPlugin;
  mainG: MultiGraph;
  closedG: MultiGraph;

  public constructor(app: App, plugin: BCPlugin) {
    this.app = app;
    this.plugin = plugin;
    this.mainG = this.plugin.mainG;
    this.closedG = this.plugin.closedG;
  }

  public DIRECTIONS = DIRECTIONS;
  public ARROW_DIRECTIONS = ARROW_DIRECTIONS;

  public buildObsGraph = () => buildObsGraph(this.app);

  public getSubInDirs = (dirs: Directions[], g = this.mainG) =>
    getSubInDirs(g, ...dirs);

  public getSubForFields = (fields: string[], g = this.mainG) =>
    getSubForFields(g, fields);

  public dfsAllPaths = (
    fromNode = this.app.workspace.getActiveFile()?.basename,
    g = this.mainG
  ) => dfsAllPaths(g, fromNode);

  public createIndex = (allPaths: string[][], wikilinks = false) =>
    createIndex(allPaths, wikilinks);

  public getMatrixNeighbours = (
    fromNode = this.app.workspace.getActiveFile()?.basename
  ) => getMatrixNeighbours(this.plugin, fromNode);

  public getOppDir = (dir: Directions) => getOppDir(dir);

  public getOppFields = (field: string) => {
    const { fieldDir } = getFieldInfo(this.plugin.settings.userHiers, field);
    return getOppFields(this.plugin.settings.userHiers, field, fieldDir);
  };

  public getFieldInfo = (field: string) =>
    getFieldInfo(this.plugin.settings.userHiers, field);
  public getFields = (dir?: Directions) =>
    getFields(this.plugin.settings.userHiers, dir ?? "all");

  public iterateHiers(
    cb: (hier: UserHier, dir: Directions, field: string) => void
  ) {
    iterateHiers(this.plugin.settings.userHiers, cb);
  }
}
