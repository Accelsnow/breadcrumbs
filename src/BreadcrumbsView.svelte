<script lang="ts">
  export let levels: string[];
</script>

<div class="breadcrumb-trail">
  {#each levels as crumb}
    <span class="crumb">{crumb}</span>
  {/each}
</div>
