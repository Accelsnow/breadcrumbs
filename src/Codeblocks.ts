import type { EdgeDefinition, NodeSingular } from "cytoscape";
import type { MultiGraph } from "graphology";
import {
  DataStoreEvents,
  getPlugin,
  ICoreDataStore,
  IJuggl,
  IJugglPlugin,
  IJugglSettings,
  IJugglStores,
  nodeDangling,
  nodeFromFile,
  VizId,
} from "juggl-api";
import { info } from "loglevel";
import {
  Component,
  Events,
  MarkdownPostProcessorContext,
  MetadataCache,
  Notice,
  TFile,
} from "obsidian";
import CBTree from "./Components/CBTree.svelte";
import {
  CODEBLOCK_FIELDS,
  CODEBLOCK_TYPES,
  DIRECTIONS,
  JUGGL_CB_DEFAULTS,
} from "./constants";
import { createIndex } from "./Commands/CreateIndex";
import {
  dfsAllPaths,
  getFieldInfo,
  getOppDir,
  getReflexiveClosure,
  getSubInDirs,
} from "./graphUtils";
import type { CodeblockFields, ParsedCodeblock } from "./interfaces";
import type BCPlugin from "./main";
import { dropFolder, getFields, splitAndTrim } from "./sharedFunctions";

export function getCodeblockCB(plugin: BCPlugin) {
  const { settings } = plugin;
  return (
    source: string,
    el: HTMLElement,
    ctx: MarkdownPostProcessorContext
  ) => {
    const parsedSource = parseCodeBlockSource(source);
    console.log(parsedSource);
    const err = codeblockError(plugin, parsedSource);

    if (err !== "") {
      el.innerHTML = err;
      return;
    }
    let min = 0,
      max = Infinity;
    let { depth, dir, from, implied, flat } = parsedSource;
    if (depth !== undefined) {
      const minNum = parseInt(depth[0]);
      if (!isNaN(minNum)) min = minNum;
      const maxNum = parseInt(depth[1]);
      if (!isNaN(maxNum)) max = maxNum;
    }

    const currFile = plugin.app.metadataCache.getFirstLinkpathDest(
      ctx.sourcePath,
      ""
    );
    const { userHiers } = settings;
    const { basename } = currFile;

    let froms = undefined;
    if (from !== undefined) {
      try {
        const api = plugin.app.plugins.plugins.dataview?.api;
        if (api) {
          const pages = api.pagePaths(from)?.values;
          froms = pages.map(dropFolder);
        } else new Notice("Dataview must be enabled for `from` to work.");
      } catch (e) {
        new Notice(`The query "${from}" failed.`);
      }
    }

    const oppDir = getOppDir(dir);
    const sub =
      implied === "false"
        ? getSubInDirs(plugin.mainG, dir)
        : getSubInDirs(plugin.mainG, dir, oppDir);
    const closed = getReflexiveClosure(sub, userHiers);
    const subClosed = getSubInDirs(closed, dir);

    const allPaths = dfsAllPaths(subClosed, basename);
    const index = createIndex(allPaths, false);
    info({ allPaths, index });
    console.log({ allPaths, index });
    const lines = index
      .split("\n")
      .map((line) => {
        const pair = line.split("- ");
        return [flat === "true" ? "" : pair[0], pair.slice(1).join("- ")] as [
          string,
          string
        ];
      })
      .filter((pair) => pair[1] !== "");

    switch (parsedSource.type) {
      case "tree":
        new CBTree({
          target: el,
          props: {
            plugin,
            el,
            min,
            max,
            lines,
            froms,
            basename,
            ...parsedSource,
          },
        });
        break;
      case "juggl":
        createdJugglCB(
          plugin,
          el,
          parsedSource,
          lines,
          froms,
          basename,
          min,
          max
        );
        break;
    }
  };
}

function parseCodeBlockSource(source: string): ParsedCodeblock {
  const lines = source.split("\n");
  const getValue = (type: string) =>
    lines
      .find((l) => l.startsWith(`${type}:`))
      ?.split(":")?.[1]
      ?.trim();

  const results: { [field in CodeblockFields]: string | boolean | string[] } =
    {};
  CODEBLOCK_FIELDS.forEach((field) => {
    results[field] = getValue(field);
    if (results[field] === "false") {
      results[field] = false;
    }
    if (results[field] === "true") {
      results[field] = true;
    }
  });

  results.field = results.field
    ? splitAndTrim(results.field as string)
    : undefined;

  if (results.depth) {
    const match = (results.depth as string).match(/(\d*)-?(\d*)/);
    results.depth = [match[1], match[2]];
  }

  return results as unknown as ParsedCodeblock;
}

function codeblockError(plugin: BCPlugin, parsedSource: ParsedCodeblock) {
  const { dir, fields, type, title, depth, flat, content, from, implied } =
    parsedSource;
  const { userHiers } = plugin.settings;
  let err = "";

  if (!CODEBLOCK_TYPES.includes(type))
    err += `<code>type: ${type}</code> is not a valid type. It must be one of: ${CODEBLOCK_TYPES.map(
      (type) => `<code>${type}</code>`
    ).join(", ")}.</br>`;

  const validDir = DIRECTIONS.includes(dir);
  if (!validDir)
    err += `<code>dir: ${dir}</code> is not a valid direction.</br>`;

  const allFields = getFields(userHiers);
  [fields].flat()?.forEach((f) => {
    if (f !== undefined && !allFields.includes(f))
      err += `<code>fields: ${f}</code> is not a field in your hierarchies.</br>`;
  });

  if (title !== undefined && title !== "false")
    err += `<code>title: ${title}</code> is not a valid value. It has to be <code>false</code>, or leave the entire line out.</br>`;

  if (depth !== undefined && depth.every((num) => isNaN(parseInt(num))))
    err += `<code>depth: ${depth}</code> is not a valid value. It has to be a number.</br>`;

  if (flat !== undefined && flat !== "true")
    err += `<code>flat: ${flat}</code> is not a valid value. It has to be <code>true</code>, or leave the entire line out.</br>`;

  if (content !== undefined && content !== "open" && content !== "closed")
    err += `<code>content: ${content}</code> is not a valid value. It has to be <code>open</code> or <code>closed</code>, or leave the entire line out.</br>`;

  if (
    from !== undefined &&
    !plugin.app.plugins.enabledPlugins.has("dataview")
  ) {
    err += `Dataview must be enabled to use <code>from</code>.</br>`;
  }

  if (implied !== undefined && implied !== "false")
    err += `<code>implied: ${implied}</code> is not a valid value. It has to be <code>false</code>, or leave the entire line out.</br>`;

  return err === ""
    ? ""
    : `${err}</br>
    A valid example would be:
    <pre><code>
      type: tree
      dir: ${validDir ? dir : "down"}
      fields: ${
        allFields
          .map((f) => {
            return { f, dir: getFieldInfo(userHiers, f).fieldDir };
          })
          .filter((info) => info.dir === dir)
          .map((info) => info.f)
          .join(", ") || "child"
      }
      depth: 3
      </code></pre>`;
}

const STORE_ID = "core";

function indentToDepth(indent: string) {
  return indent.length / 2 + 1;
}

function meetsConditions(
  indent: string,
  node: string,
  froms: string[],
  min: number,
  max: number
) {
  const depth = indentToDepth(indent);
  return (
    depth >= min &&
    depth <= max &&
    (froms === undefined || froms.includes(node))
  );
}

class BCStoreEvents extends Events implements DataStoreEvents {}

class BCStore extends Component implements ICoreDataStore {
  graph: MultiGraph;
  cache: MetadataCache;
  plugin: IJugglPlugin;
  constructor(
    graph: MultiGraph,
    metadata: MetadataCache,
    plugin: IJugglPlugin
  ) {
    super();
    this.graph = graph;
    this.cache = metadata;
    this.plugin = plugin;
  }

  asString(node: NodeSingular): string {
    const id = VizId.fromNode(node);
    return id.id.slice(0, -3);
  }

  getFile(nodeId: VizId): TFile {
    return this.cache.getFirstLinkpathDest(nodeId.id, "");
  }

  async connectNodes(
    allNodes: cytoscape.NodeCollection,
    newNodes: cytoscape.NodeCollection,
    graph: IJuggl
  ): Promise<cytoscape.EdgeDefinition[]> {
    const edges: EdgeDefinition[] = [];
    const nodesListS = new Set(
      allNodes.map((node) => this.asString(node)).filter((s) => s)
    );
    newNodes.forEach((node) => {
      console.log({ node });
      this.graph.forEachOutEdge(
        this.asString(node),
        (key, attr, source, target) => {
          if (nodesListS.has(target)) {
            edges.push({
              data: {
                id: `BC:${source}->${target}`,
                source: VizId.toId(source, STORE_ID) + ".md",
                target: VizId.toId(target, STORE_ID) + ".md",
                type: attr.field,
                dir: attr.dir,
              },
              classes: `type-${attr.field} dir-${attr.dir} breadcrumbs$`,
            });
          }
        }
      );
    });
    return Promise.resolve(edges);
  }

  getEvents(view: IJuggl): DataStoreEvents {
    return new BCStoreEvents();
  }

  getNeighbourhood(nodeId: VizId[]): Promise<cytoscape.NodeDefinition[]> {
    // TODO
    return Promise.resolve([]);
  }

  refreshNode(id: VizId, view: IJuggl): void | Promise<void> {
    return;
  }

  storeId(): string {
    return STORE_ID;
  }

  get(nodeId: VizId, view: IJuggl): Promise<cytoscape.NodeDefinition> {
    const file = this.getFile(nodeId);
    if (file === null) {
      const dangling = nodeDangling(nodeId.id);
      console.log({ dangling });
      return Promise.resolve(nodeDangling(nodeId.id));
    }
    const cache = this.cache.getFileCache(file);
    if (cache === null) {
      console.log("returning empty cache", nodeId);
      return Promise.resolve(nodeDangling(nodeId.id));
    }
    return Promise.resolve(nodeFromFile(file, this.plugin, view.settings, nodeId.toId()));
  }
}

function createJuggl(
  plugin: BCPlugin,
  target: HTMLElement,
  initialNodes: string[],
  args: IJugglSettings
) {
  try {
    const jugglPlugin = getPlugin(plugin.app);
    if (!jugglPlugin) {
      // TODO: Error handling
      return;
    }
    for (let key in JUGGL_CB_DEFAULTS) {
      if (key in args && args[key] === undefined) {
        args[key] = JUGGL_CB_DEFAULTS[key];
      }
    }

    const bcStore = new BCStore(
      plugin.mainG,
      plugin.app.metadataCache,
      jugglPlugin
    );
    const stores: IJugglStores = {
      coreStore: bcStore,
      dataStores: [bcStore],
    };

    console.log({ args }, { initialNodes });
    const juggl = jugglPlugin.createJuggl(target, args, stores, initialNodes);
    plugin.addChild(juggl);
    juggl.load();
    console.log({ juggl });
  } catch (error) {
    console.log({ error });
  }
}

export function createJugglTrail(
  plugin: BCPlugin,
  target: HTMLElement,
  paths: string[][],
  source: string,
  args: IJugglSettings
) {
  let nodes = Array.from(
    new Set(paths.reduce((prev, curr) => prev.concat(curr), []))
  );
  nodes.push(source);
  nodes = nodes.map((s) => s + ".md");
  createJuggl(plugin, target, nodes, args);
}

export function createdJugglCB(
  plugin: BCPlugin,
  target: HTMLElement,
  args: ParsedCodeblock,
  lines: [string, string][],
  froms: string[],
  source: string,
  min: number,
  max: number
) {
  const nodes = lines
    .filter(([indent, node]) => meetsConditions(indent, node, froms, min, max))
    .map(([_, node]) => node + ".md");
  if (min <= 0) {
    nodes.push(source + ".md");
  }
  console.log({ lines, nodes });
  createJuggl(plugin, target, nodes, args);
}
