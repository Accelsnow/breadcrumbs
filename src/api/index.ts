import { ListIndex } from "src/commands/list_index";
import {
	ARROW_DIRECTIONS,
	DIRECTIONS,
	type Direction,
} from "src/const/hierarchies";
import type { BCEdgeAttributes } from "src/graph/MyMultiGraph";
import { Traverse } from "src/graph/traverse";
import type { Hierarchy } from "src/interfaces/hierarchies";
import { active_file_store } from "src/stores/active_file";
import {
	get_field_hierarchy,
	get_opposite_direction,
} from "src/utils/hierarchies";
import { get } from "svelte/store";
import type BCPlugin from "../main";
import { has_edge_attrs, type EdgeAttrFilters } from "src/graph/utils";

export class BCAPI {
	plugin: BCPlugin;

	public constructor(plugin: BCPlugin) {
		this.plugin = plugin;
	}

	public DIRECTIONS = DIRECTIONS;
	public ARROW_DIRECTIONS = ARROW_DIRECTIONS;

	// TODO
	public buildObsGraph = () => {
		throw new Error("Not implemented");
	};

	public refresh = () => this.plugin.refresh();
	/** @deprecated Use refresh */
	public refreshIndex = this.refresh;

	/** @deprecated Filter edges of plugin.graph instead */
	public getSubInDirs = (dirs: Direction[], g = this.plugin.graph) => {};

	/** @deprecated Filter edges of plugin.graph instead */
	public getSubForFields = (fields: string[], g = this.plugin.graph) => {};

	public all_paths_depth_first = (
		start_node = get(active_file_store)?.path,
		graph = this.plugin.graph,
		attrs?: EdgeAttrFilters,
	) => {
		if (!start_node) throw new Error("No active file");

		return Traverse.all_paths("depth_first", graph, start_node, (e) =>
			has_edge_attrs(e, attrs),
		);
	};

	/** @deprecated Use all_paths_depth_first instead */
	public dfsAllPaths = this.all_paths_depth_first;

	public nest_all_paths = Traverse.nest_all_paths;
	public flatten_all_paths = Traverse.flatten_all_paths;

	public create_list_index = (
		start_node = get(active_file_store)?.path,
		options?: ListIndex.Options,
	) => {
		if (!start_node) throw new Error("No active file");

		return ListIndex.build(
			this.plugin.graph,
			start_node,
			Object.assign({ ...ListIndex.DEFAULT_OPTIONS }, options),
		);
	};

	// BREAKING
	/** @deprecated Use flatten_all_paths and flat_paths_to_index_list instead. Or, create_list_index */
	public createIndex = () => {};

	public get_neighbours = (source = get(active_file_store)?.path) =>
		source && this.plugin.graph.hasNode(source)
			? this.plugin.graph.get_out_edges(source)
			: [];

	/** @deprecated Use get_neighbours instead */
	public getMatrixNeighbours = this.get_neighbours;

	public get_opposite_direction = get_opposite_direction;
	/** @deprecated Use get_opposite_direction */
	public getOppDir = this.get_opposite_direction;

	public get_opposite_fields = (field: string) => {
		const field_hierarchy = get_field_hierarchy(
			this.plugin.settings.hierarchies,
			field,
		);
		if (!field_hierarchy) {
			console.error(
				"BCAPI.getOppFields: field_hierarchy not found",
				field,
			);
			return [];
		}

		return this.plugin.settings.hierarchies[field_hierarchy.hierarchy_i]
			.dirs[get_opposite_direction(field_hierarchy.dir)];
	};

	/** @deprecated Use get_opposite_fields */
	public getOppFields = this.get_opposite_fields;

	public get_field_hierarchy = (field: string) =>
		get_field_hierarchy(this.plugin.settings.hierarchies, field);

	/** @deprecated Use get_field_hierarchy */
	public getFieldInfo = this.get_field_hierarchy;

	/** @deprecated Filter plugin.settings.hierarchies instead */
	public getFields = (dir: Direction) => {};

	/** @deprecated Map plugin.settings.hierachies instead */
	public iterateHiers(
		cb: (hier: Hierarchy, dir: Direction, field: string) => void,
	) {}
}
