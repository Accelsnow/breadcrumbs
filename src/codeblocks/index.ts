import type { MarkdownPostProcessorContext } from "obsidian";
import { DIRECTIONS, type Direction } from "src/const/hierarchies";
import type { ICodeblock } from "src/interfaces/codeblocks";
import type { BreadcrumbsError } from "src/interfaces/graph";
import type BreadcrumbsPlugin from "src/main";
import { active_file_store } from "src/stores/active_file";
import { get_all_hierarchy_fields } from "src/utils/hierarchies";
import { get } from "svelte/store";
import CodeblockTree from "../components/codeblocks/CodeblockTree.svelte";

const FIELDS = [
	"type",
	"dir",
	"title",
	"fields",
	"depth",
	"flat",
	"from",
	"content",
	"sort",
] as const;

const parse_source = (plugin: BreadcrumbsPlugin, source: string) => {
	const hierarchy_fields = get_all_hierarchy_fields(
		plugin.settings.hierarchies,
	);

	const lines = source.split("\n");

	const errors: BreadcrumbsError[] = [];
	const parsed: Partial<ICodeblock["Options"]> = {};

	lines.forEach((line) => {
		const [key, ...rest] = line.split(":");
		const value = rest.join(":").trim();
		if (!key || !value) return;

		// By the time we parse the value in the switch statement,
		//   we know it's non-empty
		switch (key) {
			case "type": {
				return (parsed.type = value as ICodeblock["Options"]["type"]);
			}

			case "dir": {
				if (!DIRECTIONS.includes(value as Direction)) {
					return errors.push({
						code: "invalid_field_value",
						message: `Invalid dir: ${value}`,
						path: key,
					});
				}

				return (parsed.dir = value as Direction);
			}

			case "title": {
				return (parsed.title = value);
			}

			case "fields": {
				return (parsed.fields = value
					.split(",")
					.map((field) => field.trim())
					.filter((field) => {
						if (!hierarchy_fields.includes(field)) {
							errors.push({
								code: "invalid_field_value",
								message: `Invalid field: ${field}`,
								path: key,
							});
							return false;
						} else return true;
					}));
			}

			case "depth": {
				// depth: -2
				// depth: 1-2
				// depth: 1-

				const bounds = value.split("-").map((num) => parseInt(num));

				const [min, max] = [
					Number.isNaN(bounds[0]) ? 0 : bounds[0],
					Number.isNaN(bounds[1]) ? Infinity : bounds[1],
				];

				if (min > max) {
					errors.push({
						code: "invalid_field_value",
						message: `Invalid depth: ${value}. Min is greater than max.`,
						path: key,
					});

					return (parsed.depth = [0, Infinity]);
				}

				return (parsed.depth = [min, max] as [number, number]);
			}

			case "flat": {
				return (parsed.flat = value === "true");
			}

			case "from": {
				return (parsed.from = value);
			}

			case "content": {
				if (value !== "open" && value !== "closed") {
					return errors.push({
						code: "invalid_field_value",
						message: `Invalid content: ${value}`,
						path: key,
					});
				}

				return (parsed.content = value);
			}

			case "sort": {
				const [sort_by, sort_order] = value.split(" ");
				if (sort_by !== "basename") {
					return errors.push({
						code: "invalid_field_value",
						message: `Invalid sort_by: ${sort_by}`,
						path: key,
					});
				}

				if (sort_order !== "asc" && sort_order !== "desc") {
					return errors.push({
						code: "invalid_field_value",
						message: `Invalid sort_order: ${sort_order}`,
						path: key,
					});
				}

				parsed.sort_by = sort_by;
				parsed.sort_order = sort_order === "asc" ? 1 : -1;

				return;
			}

			default: {
				errors.push({
					code: "invalid_field_value",
					message: `Invalid field: ${key}`,
					path: key,
				});
			}
		}
	});

	return { parsed, errors };
};

const get_callback = (plugin: BreadcrumbsPlugin) => {
	return (
		source: string,
		el: HTMLElement,
		ctx: MarkdownPostProcessorContext,
	) => {
		const active_file = get(active_file_store);
		if (!active_file) return;

		const { parsed, errors } = parse_source(plugin, source);
		if (errors.length) console.log("codeblock errors", errors);

		const options: ICodeblock["Options"] = Object.assign(
			{
				type: "tree",
				dir: "down",
				depth: [0, Infinity],
				flat: false,
				// I think this is something to do with the Traverse order?
				sort_by: "default",
				sort_order: 1,
			},
			parsed,
		);
		console.log("resolved codeblock options", options);

		new CodeblockTree({
			target: el,
			props: {
				plugin,
				options,
				errors,
			},
		});
	};
};

export const Codeblocks = {
	FIELDS,
	get_callback,
};
