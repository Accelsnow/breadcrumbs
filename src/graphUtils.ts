import { MultiGraph } from "graphology";
import type { Attributes } from "graphology-types";
import type { App } from "obsidian";
// import { DIRECTIONS } from "./constants";
import type { Directions, UserHier } from "./interfaces";

// TODO - this is a hack to get the graph to work with the approvals
// I shouldn't need
const DIRECTIONS = ["up", "same", "down", "next", "prev"];
// This function takes the real & implied graphs for a given relation, and returns a new graphs with both.
// It makes implied relations real
// TODO use reflexiveClosure instead
export function closeImpliedLinks(
  real: MultiGraph,
  implied: MultiGraph
): MultiGraph {
  const closedG = real.copy();
  implied.forEachEdge((key, a, s, t) => {
    closedG.mergeEdge(t, s, a);
  });
  return closedG;
}

export function currFile(app: App) {
  return app.vault.getMarkdownFiles();
}

export function removeUnlinkedNodes(g: MultiGraph) {
  const copy = g.copy();
  copy.forEachNode((node) => {
    if (!copy.degree(node)) copy.dropNode(node);
  });
  return copy;
}

/**
 * Return a subgraph of all nodes & edges with `dirs.includes(a.dir)`
 * @param  {MultiGraph} main
 * @param  {Directions} dir
 */
export function getSubInDirs(main: MultiGraph, ...dirs: Directions[]) {
  const sub = new MultiGraph();
  main.forEachEdge((k, a, s, t) => {
    if (dirs.includes(a.dir)) {
      //@ts-ignore
      addNodesIfNot(sub, [s, t], a);
      sub.addEdge(s, t, a);
    }
  });
  return sub;
}

/**
 * Return a subgraph of all nodes & edges with `files.includes(a.field)`
 * @param  {MultiGraph} main
 * @param  {string[]} fields
 */
export function getSubForFields(main: MultiGraph, fields: string[]) {
  const sub = new MultiGraph();
  main.forEachEdge((k, a, s, t) => {
    if (fields.includes(a.field)) {
      //@ts-ignore
      addNodesIfNot(sub, [s, t], a);
      sub.addEdge(s, t, a);
    }
  });
  return sub;
}

/**
 * For every edge in `g`, add the reverse of the edge to a copy of `g`.
 *
 * It also sets the attrs of the reverse edges to `oppDir` and `oppFields[0]`
 * @param  {MultiGraph} g
 * @param  {UserHier[]} userHiers
 * @param  {boolean} closeAsOpposite
 */
export function getReflexiveClosure(
  g: MultiGraph,
  userHiers: UserHier[],
  closeAsOpposite: boolean = true
): MultiGraph {
  const copy = g.copy();
  copy.forEachEdge((k, a, s, t) => {
    const { dir, field } = a;
    if (field === undefined) return;
    const oppDir = getOppDir(dir);
    const oppField = getOppFields(userHiers, field)[0];

    addNodesIfNot(copy, [s, t], {
      //@ts-ignore
      dir: closeAsOpposite ? oppDir : dir,
      field: closeAsOpposite ? oppField : field,
    });
    addEdgeIfNot(copy, t, s, {
      //@ts-ignore
      dir: closeAsOpposite ? oppDir : dir,
      field: closeAsOpposite ? oppField : field,
    });
  });
  return copy;
}

export function addNodesIfNot(
  g: MultiGraph,
  nodes: string[],
  attr?: { order: number }
) {
  nodes.forEach((node) => {
    if (!g.hasNode(node)) g.addNode(node, attr);
  });
}

export function addEdgeIfNot(
  g: MultiGraph,
  source: string,
  target: string,
  attr?: Attributes
) {
  if (!g.hasEdge(source, target)) g.addEdge(source, target, attr);
}

export const getSinks = (g: MultiGraph) =>
  g.filterNodes((node) => g.hasNode(node) && !g.outDegree(node));

export const getSources = (g: MultiGraph) =>
  g.filterNodes((node) => g.hasNode(node) && !g.inDegree(node));

export const getOutNeighbours = (g: MultiGraph, node: string): string[] =>
  g.hasNode(node) ? g.outNeighbors(node) : [];
export const getInNeighbours = (g: MultiGraph, node: string): string[] =>
  g.hasNode(node) ? g.inNeighbors(node) : [];

export const getOppDir = (dir: Directions): Directions => {
  switch (dir) {
    case "up":
      return "down";
    case "down":
      return "up";
    case "same":
      return "same";
    case "next":
      return "prev";
    case "prev":
      return "next";
  }
};

/**
 *  Get the hierarchy and direction that `field` is in
 * */
export function getFieldInfo(userHiers: UserHier[], field: string) {
  let fieldDir: Directions;
  let fieldHier: UserHier;

  DIRECTIONS.forEach((dir: Directions) => {
    userHiers.forEach((hier) => {
      if (hier[dir].includes(field)) {
        fieldDir = dir;
        fieldHier = hier;
        return;
      }
    });
  });
  return { fieldHier, fieldDir };
}

export function getOppFields(userHiers: UserHier[], field: string) {
  const { fieldHier, fieldDir } = getFieldInfo(userHiers, field);
  const oppDir = getOppDir(fieldDir);
  return fieldHier[oppDir];
}

export function dfsAllPaths(g: MultiGraph, startNode: string): string[][] {
  const queue: { node: string; path: string[] }[] = [
    { node: startNode, path: [] },
  ];
  const visited = [];
  const allPaths: string[][] = [];

  let i = 0;
  while (queue.length > 0 && i < 1000) {
    i++;
    const { node, path } = queue.shift();

    const extPath = [node, ...path];
    const succsNotVisited = g.hasNode(node)
      ? g.filterOutNeighbors(node, (n, a) => !visited.includes(n))
      : [];
    const newItems = succsNotVisited.map((n) => {
      return { node: n, path: extPath };
    });

    visited.push(...succsNotVisited);
    queue.unshift(...newItems);

    if (!g.hasNode(node) || !g.outDegree(node)) allPaths.push(extPath);
  }
  return allPaths;
}
