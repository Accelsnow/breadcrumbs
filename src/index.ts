export { buildObsGraph } from "./Utils/graphUtils";
