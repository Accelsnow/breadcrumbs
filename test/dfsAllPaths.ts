import { MultiGraph } from "graphology";
import { dfsAllPaths } from "../src/graphUtils";
import { testGraph } from "./testUtils";

require("approvals").mocha();

function verify(sol: any) {
  this.verifyAsJSON(sol, {
    reporters: ["tortoisemerge"],
    appendEOL: true,
    normalizeLineEndingsTo: "\r\n",
  });
}

describe("dfsAllPaths", function () {
  it("1", function () {
    const g = new MultiGraph();
    g.addNode("A");
    g.addNode("B");

    g.addEdge("A", "B", { dir: "up", field: "up" });

    const paths = dfsAllPaths(g, "A");

    verify.call(this, paths);
  });

  it("2", function () {
    const g = testGraph();
    const paths = dfsAllPaths(g, "a");

    verify.call(this, paths);
  });
});

export {};
