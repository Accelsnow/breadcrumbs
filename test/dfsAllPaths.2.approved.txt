[
  [
    "f",
    "e",
    "d",
    "c",
    "b",
    "a"
  ],
  [
    "f",
    "d",
    "c",
    "b",
    "a"
  ],
  [
    "f",
    "e",
    "d",
    "b",
    "a"
  ],
  [
    "f",
    "e",
    "d",
    "c",
    "a"
  ],
  [
    "f",
    "d",
    "b",
    "a"
  ],
  [
    "f",
    "d",
    "c",
    "a"
  ]
]
